### PR TITLE
Library remainder

### PR DESCRIPTION
**Pull request by neurogoo**
Current idea is, when sendRemainder is launched, it checks which loans are 30 days and over and have loantime `mod` 30 == 0. This is so that it doens't have to keep track which was the last time it send remainders to not spam people too much. Idea would be that this is launched once everyday.

If this sounds good then, next question is what next? I probably need to at this as Service? And what kind of authentication should I put here?

### CODE REVIEW

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+import qualified Personio as P
+
+getAllOldLoans :: (MonadLog m, MonadBaseControl IO m, MonadCatch m) => Ctx -> m [LoanData]
+getAllOldLoans ctx = do
+    now <- currentDay
+    loans <- runLogT "loan-remainder" (ctxLogger ctx) $ fetchAllLoans (ctxPostgres ctx)
+    pure $ filter (oldLoansWithDayInterval now) $ loans
+  where
+      oldLoansWithDayInterval n (LoanData _ loanDay _ _) = diffDays n loanDay >= 30 && diffDays n loanDay `mod` 30 == 0
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: rename `n` and `now` to `today`

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+reminderTemplate :: Text -> Integer -> Text -> Text
+reminderTemplate loanerName days bookName = intercalate "\n"
+    ["Hello " <> loanerName <> "!"
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: space after `[`

last `]` on own line

File: library-app/src/Futurice/App/Library.hs [hunk 3]
```diff
@@ -457,6 +458,12 @@ addItemPostImpl ctx req = do
       (True, AddItemRequest _ _ (Just boardgameinformationId)) -> editBoardGameInformationPageImpl ctx boardgameinformationId
       _ -> throwError err404
 
+sendReminderEmailsImpl :: Ctx -> Maybe Login -> Handler Bool
+sendReminderEmailsImpl ctx login = withAuthUser ctx login $ (\l -> do
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: parens aren't required

File: library-app/src/Futurice/App/Library/API.hs [hunk 4]
```diff
@@ -22,17 +22,18 @@ import Futurice.App.Library.Types
...
+    , bookByISBNGet         :: route :- "book" :> "isbn" :> Capture "isbn" Text :> Get '[JSON] BookInformationByISBNResponse
+    , bookCoverGet          :: route :- BookCoverEndpoint
+    , borrowPost            :: route :- SSOUser :> "book" :> "borrow" :> ReqBody '[JSON] BorrowRequest :> Post '[JSON] Loan
+    , itemDelete            :: route :- "item"  :> Capture "id" ItemId :> Delete '[JSON] Text
+    , snatchPost            :: route :- SSOUser :> "book" :> "snatch" :> Capture "id" ItemId :> Post '[JSON] Loan
+    , loansGet              :: route :- "loan" :> Get '[JSON] [Loan]
+    , loanGet               :: route :- "loan" :> Capture "id" LoanId :> Get '[JSON] Loan
+    , returnPost            :: route :- "return" :> Capture "id" LoanId :> Post '[JSON] Bool
+    , personalLoansGet      :: route :- SSOUser :> "user" :> "loan" :> Get '[JSON] [Loan]
+    , sendReminderEmailsGet :: route :- SSOUser :> "reminder" :> Get '[JSON] Bool
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/API.hs]: `Post`.

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+          Right () -> return ()
+  where
+      emailProxyUrl = (cfgEmailProxyUrl . ctxConfig) ctx
+      manager = ctxManager ctx
+      req loanerId days loanerEmail iName = Req
+          { _reqTo = loanerEmail
+          , _reqCc = Nothing
+          , _reqBcc = Nothing
+          , _reqReplyTo = Nothing
+          , _reqSubject = "LIBRARY: Reminder"
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: can this include book name?

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+sendReminderEmails ctx emps = do
+    now <- currentDay
+    oldLoans <- getAllOldLoans ctx
+    reqs <- for oldLoans $ \l -> runMaybeT $ do
+      loaner <- MaybeT $ pure $ emps ^.at (ldPersonioId l)
+      loanerEmail <- MaybeT $ pure $ loaner ^. P.employeeEmail
+      itemName <- do
+        itemData <- MaybeT $ fetchItem (ctxPostgres ctx) (ldItemId l)
+        case idInfoId itemData of
+          BookInfoId bookInfoid -> (fmap . fmap) (^. bookTitle) (fetchBookInformation (ctxPostgres ctx) bookInfoid)
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: please use `view bookTitle` when partially applied. `(^. bookTitle) is ugly.

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+getAllOldLoans :: (MonadLog m, MonadBaseControl IO m, MonadCatch m) => Ctx -> m [LoanData]
+getAllOldLoans ctx = do
+    now <- currentDay
+    loans <- runLogT "loan-remainder" (ctxLogger ctx) $ fetchAllLoans (ctxPostgres ctx)
+    pure $ filter (oldLoansWithDayInterval now) $ loans
+  where
+      oldLoansWithDayInterval n (LoanData _ loanDay _ _) = diffDays n loanDay >= 30 && diffDays n loanDay `mod` 30 == 0
+
+reminderTemplate :: Text -> Integer -> Text -> Text
+reminderTemplate loanerName days bookName = intercalate "\n"
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: `unlines`

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import Prelude ()
+
+import Futurice.App.Library.Config
+import Futurice.App.Library.Ctx
+import Futurice.App.Library.Logic
+import Futurice.App.Library.Types
+
+import qualified Personio as P
+
+getAllOldLoans :: (MonadLog m, MonadBaseControl IO m, MonadCatch m) => Ctx -> m [LoanData]
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: `MonadLog` yet, `runLogT`? I don't understand.

File: library-app/src/Futurice/App/Library.hs (hunk 3 above)
Comment (phadej) [reply library-app/src/Futurice/App/Library.hs]: This should guard to allow only say, `TeamIT`.

You plan to connect this to Cloudwatch Events, don't you?

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    ["Hello " <> loanerName <> "!"
+    , "You've had the following book loaned for more than "<> pack (show days) <> " days"
+    , "- " <> bookName
+    , "Please return the book if you are not reading it anymore"
+    , "https://library.app.futurice.com/user/page"
+    , "~Friendly Futurice Library Assistant"]
+
+sendReminderEmails :: (MonadIO m, MonadLog m, MonadBaseControl IO m, MonadCatch m) => Ctx -> IdMap P.Employee ->  m ()
+sendReminderEmails ctx emps = do
+    now <- currentDay
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: `now` -> `today`

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  where
+      oldLoansWithDayInterval n (LoanData _ loanDay _ _) = diffDays n loanDay >= 30 && diffDays n loanDay `mod` 30 == 0
+
+reminderTemplate :: Text -> Integer -> Text -> Text
+reminderTemplate loanerName days bookName = intercalate "\n"
+    ["Hello " <> loanerName <> "!"
+    , "You've had the following book loaned for more than "<> pack (show days) <> " days"
+    , "- " <> bookName
+    , "Please return the book if you are not reading it anymore"
+    , "https://library.app.futurice.com/user/page"
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: don't hardcode the public url

Comment (phadej) [reply library-app/src/Futurice/App/Library/Reminder.hs]: and, use `safeLink` functionality, i.e. don't hardcode path either

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    now <- currentDay
+    loans <- runLogT "loan-remainder" (ctxLogger ctx) $ fetchAllLoans (ctxPostgres ctx)
+    pure $ filter (oldLoansWithDayInterval now) $ loans
+  where
+      oldLoansWithDayInterval n (LoanData _ loanDay _ _) = diffDays n loanDay >= 30 && diffDays n loanDay `mod` 30 == 0
+
+reminderTemplate :: Text -> Integer -> Text -> Text
+reminderTemplate loanerName days bookName = intercalate "\n"
+    ["Hello " <> loanerName <> "!"
+    , "You've had the following book loaned for more than "<> pack (show days) <> " days"
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: `textShow`

File: library-app/src/Futurice/App/Library/Reminder.hs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import Futurice.App.Library.Ctx
+import Futurice.App.Library.Logic
+import Futurice.App.Library.Types
+
+import qualified Personio as P
+
+getAllOldLoans :: (MonadLog m, MonadBaseControl IO m, MonadCatch m) => Ctx -> m [LoanData]
+getAllOldLoans ctx = do
+    now <- currentDay
+    loans <- runLogT "loan-remainder" (ctxLogger ctx) $ fetchAllLoans (ctxPostgres ctx)
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Reminder.hs]: `ctxPostgres ctx` -> `ctx`.

There should be

```haskell
instance HasPostgresPool Ctx where
    postgresPool = ctxPostgres
```

I.e. `fetchAllLoans` should be

```haskell
fetchAllLoans :: (..., HasPostgresPool ctx) => ctx -> ...
```

Comment (phadej) [reply library-app/src/Futurice/App/Library/Reminder.hs]: `runLogT` is probably unnecessary

File: library-app/src/Futurice/App/Library/Reminder.hs (hunk 5 above)
Comment (neurogoo) [reply library-app/src/Futurice/App/Library/Reminder.hs]: I would leave the book name out as that can be somewhat long. And also, if they need to know which books are reminded of maybe they have too many books loaned

File: library-app/src/Futurice/App/Library/Reminder.hs (hunk 8 above)
Comment (neurogoo) [reply library-app/src/Futurice/App/Library/Reminder.hs]: I started without MonadLog and forgot to remove this after adding it

File: library-app/src/Futurice/App/Library/Reminder.hs (hunk 5 above)
Comment (phadej) [reply library-app/src/Futurice/App/Library/Reminder.hs]: Some changes would be nice so this doesn't look like complete scam/spam :)

Reminder of what?

Comment (neurogoo) [reply library-app/src/Futurice/App/Library/Reminder.hs]: Well does `LIBRARY: Book loan reminder` sound better?

File: library-app/src/Futurice/App/Library.hs (hunk 3 above)
Comment (neurogoo) [reply library-app/src/Futurice/App/Library.hs]: Sure though I don't remember all the things that is necessary to change to get that happen. But does Cloudwatch have TeamIT priviledges?

Comment (neurogoo) [reply library-app/src/Futurice/App/Library.hs]: Actually I should probably remove this API and just have forEachMessage running listening to calls?

File: library-app/src/Futurice/App/Library/Reminder.hs (hunk 5 above)
Comment (phadej) [reply library-app/src/Futurice/App/Library/Reminder.hs]: yeah. Maybe `[Library] Bool loan reminder` to SCREAM LESS

File: library-app/src/Futurice/App/Library.hs (hunk 3 above)
Comment (phadej) [reply library-app/src/Futurice/App/Library.hs]: The api endpoint might be useful as a "force button". But not necessary.

File: library-app/src/Futurice/App/Library.hs [hunk 13]
```diff
@@ -451,15 +464,39 @@ itemDeleteImpl ctx itemid = runLogT "delete-item" (ctxLogger ctx) $ do
...
+    return True
+
 withAuthUser :: Ctx -> Maybe Login -> (Login -> Handler a) -> Handler a
 withAuthUser ctx loc f = case loc <|> cfgMockUser cfg of
     Nothing    -> throwError err403
     Just login -> f login
   where
     cfg = ctxConfig ctx
+
+withAuthUser'
```
Comment (phadej) [inline library-app/src/Futurice/App/Library.hs]: :+1:

File: library-app/src/Futurice/App/Library/Pages/PersonalLoansPage.hs [hunk 14]
```diff
@@ -14,7 +14,10 @@ import qualified Data.Text as T
 personalLoansPage :: [Loan] -> HtmlPage ("personalinformation")
 personalLoansPage loans = page_ "My loans" (Just NavUser) $ do
     let (bookLoans, boardgameLoans) = loanSorter loans
-    unless (null bookLoans && null boardgameLoans) $ fullRow_ $ do
+    if null bookLoans && null boardgameLoans then
```
Comment (phadej) [inline library-app/src/Futurice/App/Library/Pages/PersonalLoansPage.hs]: put the `then` on the next line

**Pull request merged**